File: apps/web/src/types/source-checks.ts
export type SourceCheckRun = {
  id: string;
  workspace_id: string | null;
  status: string;
  checked_query_count: number;
  result_count: number;
  error: string | null;
  started_at: string;
  finished_at: string | null;
};

export type SourceCheckResult = {
  id: string;
  run_id: string;
  tracking_query_id: string;
  title: string;
  url: string | null;
  snippet: string | null;
  source_name: string | null;
  source_hint: string | null;
  raw: Record<string, unknown> | null;
  checked_at: string;
};

export type SourceCheckRunData = {
  run: SourceCheckRun;
  results: SourceCheckResult[];
};

export type SourceCheckRunListData = {
  items: SourceCheckRun[];
  total: number;
};

export type SourceProviderHealth = {
  source_hint: string;
  enabled_query_count: number;
  recent_result_count: number;
  last_result_at: string | null;
  recent_error_count: number;
  latest_error: string | null;
  recent_errors: {
    tracking_query_id: string;
    query: string;
    error: string;
  }[];
  latest_run_status: string | null;
  latest_run_error: string | null;
};

export type SourceProviderHealthData = {
  items: SourceProviderHealth[];
  total: number;
};

File: backend/app/schemas/source_checks.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SourceCheckRunRequest(BaseModel):
    limit: int = Field(20, ge=1, le=100)


class SourceCheckResultRead(BaseModel):
    id: UUID
    run_id: UUID
    tracking_query_id: UUID
    title: str
    url: str | None
    snippet: str | None
    source_name: str | None
    source_hint: str | None
    raw: dict[str, Any] | None
    checked_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SourceCheckRunRead(BaseModel):
    id: UUID
    workspace_id: UUID | None
    status: str
    checked_query_count: int
    result_count: int
    error: str | None
    started_at: datetime
    finished_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class SourceCheckRunData(BaseModel):
    run: SourceCheckRunRead
    results: list[SourceCheckResultRead]


class SourceCheckRunListData(BaseModel):
    items: list[SourceCheckRunRead]
    total: int


class SourceProviderHealthErrorRead(BaseModel):
    tracking_query_id: UUID
    query: str
    error: str


class SourceProviderHealthRead(BaseModel):
    source_hint: str
    enabled_query_count: int
    recent_result_count: int
    last_result_at: datetime | None
    recent_error_count: int
    latest_error: str | None
    recent_errors: list[SourceProviderHealthErrorRead] = Field(default_factory=list)
    latest_run_status: str | None
    latest_run_error: str | None


class SourceProviderHealthData(BaseModel):
    items: list[SourceProviderHealthRead]
    total: int
